memkind_default.3
.\"
.\" Copyright (C) 2014 Intel Corporation.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\" 1. Redistributions of source code must retain the above copyright notice(s),
.\"    this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright notice(s),
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
.\" EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.TH "MEMKIND_DEFAULT" 3 "2014-09-22" "Intel Corporation" "MEMKIND_DEFAULT" \" -*- nroff -*-
.SH "NAME"
memkind_default.h \- default implementations for memkind operations.
.SH "SYNOPSIS"
.nf
.B #include <memkind_default.h>
.sp
.B Link with -ljemalloc -lnuma -lpthread -lmemkind
.sp
.BI "int memkind_default_create(struct memkind " "*kind" ", const struct memkind_ops " "*ops" ", const char " "*name" );
.br
.BI "int memkind_default_destroy(struct memkind " "*kind" );
.br
.BI "void *memkind_default_malloc(struct memkind " "*kind" ", size_t " "size" );
.br
.BI "void *memkind_default_calloc(struct memkind " "*kind" ", size_t " "num" ", size_t " "size" );
.br
.BI "int memkind_default_posix_memalign(struct memkind " "*kind" ", void " "**memptr" ", size_t " "alignment" ", size_t " "size" );
.br
.BI "void *memkind_default_realloc(struct memkind " "*kind" ", void " "*ptr" ", size_t " "size" );
.br
.BI "void memkind_default_free(struct memkind " "*kind" ", void " "*ptr" );
.br
.BI "int memkind_default_mbind(struct memkind " "*kind" ", void " "*ptr" ", size_t " "len" );
.br
.BI "int memkind_default_get_mmap_flags(struct memkind " "*kind" ", int " "*flags" );
.br
.BI "int memkind_default_get_mbind_mode(struct memkind " "*kind" ", int " "*mode" );
.br
.BI "int memkind_preferred_get_mbind_mode(struct memkind " "*kind" ", int " "*mode" );
.br
.BI "int memkind_default_get_size(struct memkind " "*kind" ", size_t " "*total" ", size_t " "*free" );
.br
.BI "int memkind_posix_check_alignment(struct memkind " "*kind" ", size_t " "alignment" );
.br
.SH DESCRIPTION
.PP
Default implementations for memkind operations which include a several
useful methods that are not part of the
.B MEMKIND_DEFAULT
kind which is a fall through to the jemalloc implementation.
.PP
.BR memkind_default_create ()
implements the required start up for every kind.  If a kind does not
point to this function directly for its
.BR ops.create ()
operation, then the function that it points to must call
.BR memkind_default_create ()
at its start.
.PP
.BR memkind_default_destroy ()
implements the required shutdown for every kind.  If a kind does not
point to this function directly for its
.BR ops.destroy ()
operation, then the function that it points to must call
.BR memkind_default_destroy ()
at its end.
.PP
.BR memkind_default_malloc ()
is a direct call through to
.BR je_malloc ().
.PP
.BR memkind_default_calloc ()
is a direct call through to
.BR je_calloc ().
.PP
.BR memkind_default_posix_memalign ()
is a direct call through to
.BR je_posix_memalign ().
.PP
.BR memkind_default_realloc ()
is a direct call through to
.BR je_realloc ().
.PP
.BR memkind_default_free ().
is a direct call through to
.BR je_free ().
Note that this method can be called on any pointer returned by a
jemalloc allocation, and in particular, all of the arena
allocations described in
.BR memkind_arena (3)
can use this function for freeing.
.PP
.BR memkind_default_mbind ()
makes calls the kind's
.BR ops.get_mbind_nodemask ()
and
.BR ops.get_mbind_mode ()
operations to gather inputs and then calls the
.BR mbind (2)
system call using the results along with and user input
.I ptr
and
.IR len .
.PP
.BR memkind_default_get_mmap_flags ()
sets
.I flags
to
.BR "MAP_PRIVATE | MAP_ANONYMOUS" .
See
.BR mmap (2)
for more information about these flags.
.PP
.BR memkind_default_get_mbind_mode ()
sets
.I mode
to
.BR MPOL_BIND .
See
.BR mbind (2)
for more information about this flag.
.PP
.BR memkind_preferred_get_mbind_mode ()
sets
.I mode
to
.BR MPOL_PREFERRED .
See
.BR mbind (2)
for more information about this flag.
.PP
.BR memkind_default_get_size ()
sums the values returned by
.BR numa_node_size64 ()
function for every NUMA node set by the
.BR .get_node_mask ()
operation.  If the 
.I .get_node_mask
pointer is 
.B NULL
then the values are summed for all NUMA nodes.
.PP
.BR memkind_posix_check_alignment ()
can be used to check the alignment value for
.BR memkind_posix_memalign ()
to ensure that is abides by the POSIX requirements:
alignment must be a power of 2 at least as large as
.RB sizeof( "void *" ).
.SH "COPYRIGHT"
Copyright (C) 2014 Intel Corporation. All rights reserved.
.SH "SEE ALSO"
.BR memkind (3),
.BR memkind_arena (3),
.BR memkind_hbw (3),
.BR memkind_hugetlb (3),
.BR memkind_gbtlb (3),
.BR jemalloc (3),
.BR mbind (2),
.BR mmap (2)


memkind_default.c
/*
 * Copyright (C) 2014 Intel Corporation.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 1. Redistributions of source code must retain the above copyright notice(s),
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice(s),
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO
 * EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <numa.h>
#include <numaif.h>
#include <sys/mman.h>
#include <errno.h>
#include <jemalloc/jemalloc.h>
#include "memkind_default.h"


int memkind_default_create(struct memkind *kind, const struct memkind_ops *ops, const char *name)
{
    int err = 0;

    kind->ops = ops;
    if (strlen(name) >= MEMKIND_NAME_LENGTH) {
        kind->name[0] = '\0';
        err = MEMKIND_ERROR_INVALID;
    }
    if (!err) {
        kind->name[MEMKIND_NAME_LENGTH-1] = '\0';
        strncpy(kind->name, name, MEMKIND_NAME_LENGTH-1);
    }
    return err;
}

int memkind_default_destroy(struct memkind *kind)
{
    kind->name[0] = '\0';
    return 0;
}

void *memkind_default_malloc(struct memkind *kind, size_t size)
{
    return je_malloc(size);
}

void *memkind_default_calloc(struct memkind *kind, size_t num, size_t size)
{
    return je_calloc(num, size);
}

int memkind_default_posix_memalign(struct memkind *kind, void **memptr, size_t alignment, size_t size)
{
    return je_posix_memalign(memptr, alignment, size);
}

void *memkind_default_realloc(struct memkind *kind, void *ptr, size_t size)
{
    return je_realloc(ptr, size);
}

void memkind_default_free(struct memkind *kind, void *ptr)
{
    je_free(ptr);
}

int memkind_default_get_size(struct memkind *kind, size_t *total, size_t *free)
{
    nodemask_t nodemask;
    struct bitmask nodemask_bm = {NUMA_NUM_NODES, nodemask.n};
    long long f;
    int err = 0;
    int i;

    *total = 0;
    *free = 0;
    if (kind->ops->get_mbind_nodemask) {
        err = kind->ops->get_mbind_nodemask(kind, nodemask.n, NUMA_NUM_NODES);
    }
    else {
        copy_bitmask_to_bitmask(numa_all_nodes_ptr, &nodemask_bm);
    }
    if (!err) {
        for (i = 0; i < NUMA_NUM_NODES; ++i) {
            if (numa_bitmask_isbitset(&nodemask_bm, i)) {
                *total += numa_node_size64(i, &f);
                *free += f;
            }
        }
    }
    return err;
}

int memkind_default_mbind(struct memkind *kind, void *ptr, size_t len)
{
    nodemask_t nodemask;
    int err = 0;
    int mode;

    err = kind->ops->get_mbind_nodemask(kind, nodemask.n, NUMA_NUM_NODES);
    if (!err) {
        err = kind->ops->get_mbind_mode(kind, &mode);
    }
    if (!err) {
        err = mbind(ptr, len, mode, nodemask.n, NUMA_NUM_NODES, 0);
        err = err ? MEMKIND_ERROR_MBIND : 0;
    }
    return err;
}

int memkind_default_get_mmap_flags(struct memkind *kind, int *flags)
{
    *flags = MAP_PRIVATE | MAP_ANONYMOUS;
    return 0;
}

int memkind_default_get_mbind_mode(struct memkind *kind, int *mode)
{
    *mode = MPOL_BIND;
    return 0;
}

int memkind_preferred_get_mbind_mode(struct memkind *kind, int *mode)
{
    *mode = MPOL_PREFERRED;
    return 0;
}

int memkind_posix_check_alignment(struct memkind *kind, size_t alignment)
{
    int err = 0;
    if ((alignment < sizeof(void*)) ||
        (((alignment - 1) & alignment) != 0)) {
            err = EINVAL;
    }
    return err;
}
